packages/model/src/projection.ts

```diff
@@ -1,9 +1,321 @@
+// tells if value is valid (meaning that it only has the fields allowed by the actual projection)
+export function respectsProjection<T extends types.Type>(
+  type: T,
+  projection: projection.Infer<T>,
+  value: types.Infer<T>,
+): boolean {
+  failWithInternalError('TODO: This is not implemented, if you catch this means we forgot an implementation!')
+}
+
+/**
+ * A record of {@link Projection `Projection`s}.
+ */
+type Projections = Record<string, types.OptionalType<Projection>>
+
+/**
+ * Given a Mondrian {@link Type type}, returns the Mondrian type describing its {@link Projection projection}.
+ * You can read {@link here TODO:} to learn more about what projections are and how they can be used.
+ *
+ * @example ```ts
+ *          const model = object({ field1: number, field2: string })
+ *          type Projection = InferProjection<typeof model>
+ *          Infer<Projection>
+ *          // -> true | { field1?: true, field2?: true }
+ *          ```
+ */
+// prettier-ignore
+export type FromType<T extends types.Type>
+  = [T] extends [types.NumberType] ? types.LiteralType<true>
+  : [T] extends [types.StringType] ? types.LiteralType<true>
+  : [T] extends [types.BooleanType] ? types.LiteralType<true>
+  : [T] extends [types.EnumType<infer _>] ? types.LiteralType<true>
+  : [T] extends [types.LiteralType<infer _>] ? types.LiteralType<true>
+  : [T] extends [types.CustomType<infer _Name, infer _Options, infer _InferredAs>] ? types.LiteralType<true>
+  : [T] extends [types.ArrayType<infer _, infer T1>] ? projection.FromType<T1>
+  : [T] extends [types.OptionalType<infer T1>] ? projection.FromType<T1>
+  : [T] extends [types.NullableType<infer T1>] ? projection.FromType<T1>
+  : [T] extends [types.ReferenceType<infer T1>] ? projection.FromType<T1>
+  : [T] extends [(() => infer T1 extends types.Type)] ? projection.FromType<T1>
+  : [T] extends [types.UnionType<infer Ts>] ? types.UnionType<{ 
+      all: types.LiteralType<true>,
+      partial: types.ObjectType<"immutable", { [Key in keyof Ts]: types.OptionalType<projection.FromType<Ts[Key]>> }>
+    }>
+  : [T] extends [types.ObjectType<infer _, infer Ts>] ? types.UnionType<{
+      all: types.LiteralType<true>,
+      partial: types.ObjectType<"immutable", { [Key in keyof Ts]: types.OptionalType<projection.FromType<Ts[Key]>> }>
+    }>
+  : never
+
+/**
+ * @param type the type whose projection model is returned
+ * @returns the Mondrian model describing the type of valid projections for the given type
+ * @example ```ts
+ *          const model = object({ field1: number, field2: string })
+ *          projectionFromType(model)
+ *          // -> union({
+ *          //   all: literal(true)
+ *          //   partial: object({
+ *          //     field1: literal(true).optional()
+ *          //     field2: literal(true).optional()
+ *          //   })
+ *          // })
+ *          ```
+ */
+export function fromType<T extends types.Type>(type: T): projection.FromType<T> {
```

Comment:
Maybe we'll use it in openapi generation. The return type can be more generic like:
```typescript
type UnknownProjection = true | { [K in string]?: true | UnknownProjection }
```